cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -866,70 +657,54 @@ def bulk_migrate_from_modulestore(
-                    if (
-                        source_key in previous_migration
-                        and previous_migration[source_key].migrations__target_collection__key
-                    ):
-                        # Has previous migration with collection
-                        try:
-                            # Get the previous collection
-                            previous_collection = authoring_api.get_collection(
-                                target_package.id,
-                                previous_migration[source_key].migrations__target_collection__key,
-                            )
-
-                            migration.target_collection = previous_collection
-                        except Collection.DoesNotExist:
-                            # The collection no longer exists or is being migrated to a different library.
-                            # In that case, create a new collection independent of strategy
-                            migration.target_collection = _create_collection(target_library_locator, title)
-                    else:
-                        # Create collection and save in migration
-                        migration.target_collection = _create_collection(target_library_locator, title)
-
+                existing_collection_to_use: Collection | None = None
+                # For Fork strategy: Create an new collection every time.
+                # For Update and Skip strategies: Update an existing collection if possible.
+                if migration.repeat_handling_strategy != RepeatHandlingStrategy.Fork.value:
+                    if source_data.previous_migration:
+                        if previous_collection_slug := source_data.previous_migration.target_collection_slug:
+                            try:
+                                existing_collection_to_use = authoring_api.get_collection(
+                                    target_package.id, previous_collection_slug
+                                )
+                            except Collection.DoesNotExist:
+                                # Collection no longer exists.
+                                pass
+                migration.target_collection = (
+                    existing_collection_to_use or
+                    _create_collection(library_key=target_library_locator, title=legacy_root_list[i].display_name)
+                )
             _populate_collection(user_id, migration)
-
-        ModulestoreMigration.objects.bulk_update(
+        models.ModulestoreMigration.objects.bulk_update(
             [x.migration for x in source_data_list],
             ["target_collection", "is_failed"],
         )
         status.increment_completed_steps()
     except Exception as exc:  # pylint: disable=broad-exception-caught
         # If there is an exception in this block, all migrations fail.
-        _set_migrations_to_fail(source_data_list)
+        log.exception("Modulestore migrations failed")
```

Comment:
We've added some exception logging.

(I deleted _set_migrations_to_fail by accident, will reinstate before merging)
```suggestion
         _set_migrations_to_fail(source_data_list)
        log.exception("Modulestore migrations failed")
```